substrate/client/rpc-spec-v2/src/chain_head/subscription/inner.rs

```diff
@@ -155,113 +162,55 @@ struct PermitOperations {
 	_permit: tokio::sync::OwnedSemaphorePermit,
 }
 
-/// The state of one operation.
-///
-/// This is directly exposed to users via `chain_head_unstable_continue` and
-/// `chain_head_unstable_stop_operation`.
+/// Stop handle for the operation.
 #[derive(Clone)]
-pub struct OperationState {
-	/// The shared operation state that holds information about the
-	/// `waitingForContinue` event and cancellation.
-	shared_state: Arc<SharedOperationState>,
-	/// Send notifications when the user calls `chainHead_continue` method.
-	send_continue: tokio::sync::mpsc::Sender<()>,
-}
+pub struct StopHandle(tokio::sync::mpsc::Sender<()>);
 
-impl OperationState {
-	/// Returns true if `chainHead_continue` is called after the
-	/// `waitingForContinue` event was emitted for the associated
-	/// operation ID.
-	pub fn submit_continue(&self) -> bool {
-		// `waitingForContinue` not generated.
-		if !self.shared_state.requested_continue.load(std::sync::atomic::Ordering::Acquire) {
-			return false
-		}
-
-		// Has enough capacity for 1 message.
-		// Can fail if the `stop_operation` propagated the stop first.
-		self.send_continue.try_send(()).is_ok()
+impl StopHandle {
+	pub async fn stopped(&self) {
+		self.0.closed().await;
 	}
 
-	/// Stops the operation if `waitingForContinue` event was emitted for the associated
-	/// operation ID.
-	///
-	/// Returns nothing in accordance with `chainHead_v1_stopOperation`.
-	pub fn stop_operation(&self) {
-		// `waitingForContinue` not generated.
-		if !self.shared_state.requested_continue.load(std::sync::atomic::Ordering::Acquire) {
-			return
-		}
-
-		self.shared_state
-			.operation_stopped
-			.store(true, std::sync::atomic::Ordering::Release);
-
-		// Send might not have enough capacity if `submit_continue` was sent first.
-		// However, the `operation_stopped` boolean was set.
-		let _ = self.send_continue.try_send(());
+	pub fn is_stopped(&self) -> bool {
+		self.0.is_closed()
 	}
 }
 
 /// The shared operation state between the backend [`RegisteredOperation`] and frontend
 /// [`RegisteredOperation`].
-struct SharedOperationState {
-	/// True if the `chainHead` generated `waitingForContinue` event.
-	requested_continue: AtomicBool,
-	/// True if the operation was cancelled by the user.
-	operation_stopped: AtomicBool,
+#[derive(Clone)]
+pub struct OperationState {
+	stop: StopHandle,
+	operations: SharedOperations,
+	operation_id: String,
 }
 
-impl SharedOperationState {
-	/// Constructs a new [`SharedOperationState`].
-	///
-	/// This is efficiently cloned under a single heap allocation.
-	fn new() -> Arc<Self> {
-		Arc::new(SharedOperationState {
-			requested_continue: AtomicBool::new(false),
-			operation_stopped: AtomicBool::new(false),
-		})
+impl OperationState {
+	pub fn stop(&mut self) {
+		if !self.stop.is_stopped() {
+			self.operations.lock().remove(&self.operation_id);
```

Comment:
annoying with to locking the mutex here instead of AtomicBool but this is needed to have an async notification when operation is stopped,

Reply:
I think this is acceptable since the `operationStop` shouldn't be happening too often if at all. We are also acquiring the mutex on dropping `RegisteredOperations` to clean up the tracing of operation IDs.